### DISCUSSION TITLE
NestJS Microservices - How to Use MikroORM in Monorepo Architecture

### DISCUSSION THREAD

**Discussion by EvilCheetah** [mikro-orm/mikro-orm#4136] [Q&A]
I am using default NestJS Monorepo Architecture in combination with Mikro-ORM.
I have no problem booting up the application using MikroORM Module, but **have a problem creating the migrations**. 
I have set up a mikro-orm config file, but have issues pulling up the data from the .env file. I have tried to instantiate a ConfigService from NestJS, but I am not able to point a file for it. A potential solution seems to be installing `dotenv`, but NestJS Config Module uses both dotenv and dotenv-expand under the hood, and doesn't seem as an elegant solution. 

My main question is:
`How to properly set up migration files in monorepo approach in NestJS ecosystem?`
OR
`Are there any recommendations on the way I tried to implement it?`

**Comment by B4nan** [comment]
Can't help much unless you provide more details, ideally a complete repro - in fact, you haven't provided any details at all. 

One thing I can say right ahead, don't expect the ORM CLI to respect your nestjs configuration, the ORM has no idea on your project dependencies like nestjs. If you want to reuse your nestjs services, you will need to write your own scripts to work with the migrations, the ORM CLI can't do that, it's a non-goal.

**Comment by LazeKer** [comment]
In the hope it'll guide you, I want to share my way of working with MikroOrm (migrations), NestJS and NX (monorepo)

I'm making the assumption that you setup a classic NestJS application and have all your entities in different libraries/modules.
The requirements of my project : 
- Respect of the MonoRepo philosophy by allowing multiple backend applications (and therefore possibly multiple databases)
- Ease of use in CI environment
- Not allowing folder scan if possible (load entities through directory search ex : /libs/**/*.entity.ts)

The solution took me sometimes to configure, but at the end it's not that complicated.

### 1. NestJS usage
Unsurprinsigly, I use the forRoot and forFeature static methods to set-up MikroOrm and load my entities automatically: 
```typescript
// app.module.ts
@Module({
  imports: [
    ConfigModule.forRoot(),
    MikroOrmModule.forRoot({
      autoLoadEntities: true,
    }),
    ApiLibraryModule,
  ],
})
export class AppModule {}
```

And in a library:
```typescript
// api-library.module.ts
@Module({
  imports: [MikroOrmModule.forFeature([Entity1, Entity2])],
})
export class ApiLibraryModule{
}
```

### 2. NX project.json scope
All my MikroOrm commands are made in a project.json scope by adding it to the commands: 
_in the library or application project.json file (see: https://nx.dev/recipes/executors/run-commands-executor)_
```json
"mikro": {
	"command": "npx mikro-orm"
},
```

And using it with NX
```bash
nx run api:mikro migration:create
```

### 3. MikroOrm configuration
Now that the command run in the context of your library/app `src` folder, MikroOrm will be able to load the .env file within it.
Mine looks something like: 
```.env
# NestJS
NEST_PORT=3000

# Mikro ORM
MIKRO_ORM_TYPE=postgresql
MIKRO_ORM_PORT=5432
MIKRO_ORM_USER=default
MIKRO_ORM_PASSWORD=default
MIKRO_ORM_DB_NAME=demo
MIKRO_ORM_HOST=localhost

# Mikro ORM CLI
MIKRO_ORM_CLI=./apps/api/mikro-orm.config.ts
MIKRO_ORM_CLI_USE_TS_NODE=true
MIKRO_ORM_CLI_TS_CONFIG_PATH=./apps/api/tsconfig.app.json
```

You probably noticed that I use the `mikro-orm.config.ts` for the rest of the configuration that is CLI only:
```ts
module.exports = async (): Promise<Options> => {
  // Used for CLI command (ex: Migration) see .env
  return <Options>{
    tsNode: true,
    entities: await loadEntities([import('@libs/api-library')]),
    migrations: {
      path: './tools/mikro-orm/api/migrations',
    },
    // See https://mikro-orm.io/docs/seeding#seeders
    seeder: {
      path: './libs/api-library/**', // Location of DatabaseSeeder
      defaultSeeder: 'DatabaseSeeder',
    },
  };
};
```

One of the main trick here is that I use the native import to load my library modules and give it to a very small helper function:
```ts
// ...
entities: await loadEntities([import('@libs/api-library')]),
// ...
```
The idea is to load the libraries modules so MikroOrm load the entities in an internal storage 
_This is done with the MikroOrmEntitiesStorage, through `forFeature` https://github.com/mikro-orm/nestjs/blob/44c680abd7211f6d768803cf25fca2420c6ef2ec/src/mikro-orm.module.ts#L41_
```ts
export async function loadEntities(
	modules: Promise<any>[],
	contextName = 'default'
): Promise<EntityName<Partial<any>>[]> {
	// Load modules
	await Promise.all(modules);

	// Retrieve entities from MikroOrmEntitiesStorage
	return [...MikroOrmEntitiesStorage.getEntities(contextName)];
}
```

### Conclusion

It may not be the intended way to implement it, but in this case (MonoRepo & Nestjs), you can customize which libraries/modules will be migrated in your applications databases and you keep a clean track/dependencies to your modules.

Notice that I didn't managed to directly provide the seeders the same way, keeping the directory search, MikroOrm not allowing us (yet?) to directly provide it.
